server/src/delta_cache.rs

```diff
@@ -0,0 +1,286 @@
+use std::collections::VecDeque;
+use unleash_types::client_features::{DeltaEvent, ClientFeature, Segment};
+
+
+
+#[derive(Debug, Clone)]
+pub struct DeltaHydrationEvent {
+    event_id: i32,
+    features: Vec<ClientFeature>,
+    segments: Vec<Segment>,
+}
+
+
+#[derive(Debug, Clone)]
+pub struct DeltaCache {
+    max_length: usize,
+    events: VecDeque<DeltaEvent>,
+    hydration_event: DeltaHydrationEvent,
+}
+
+trait GetEventId {
+    fn get_event_id(&self) -> i32;
+}
+
+impl GetEventId for DeltaEvent {
+    fn get_event_id(&self) -> i32 {
+        match self {
+            DeltaEvent::FeatureUpdated { event_id, .. }
+            | DeltaEvent::FeatureRemoved { event_id, .. }
+            | DeltaEvent::SegmentUpdated { event_id, .. }
+            | DeltaEvent::SegmentRemoved { event_id, .. }
+            | DeltaEvent::Hydration { event_id, .. } => *event_id,
+        }
+    }
+}
+
+impl DeltaCache {
+    pub fn new(hydration_event: DeltaHydrationEvent, max_length: usize) -> Self {
+        let mut cache = DeltaCache {
+            max_length,
+            events: VecDeque::new(),
+            hydration_event: hydration_event.clone(),
+        };
+        cache.add_base_event_from_hydration(hydration_event);
```

Comment:
This is more a nit than something that needs to be done (and it's good to start thinking like a Rust developer).

`add_base_event_from_hydration` currently takes an owned type but then clones the stuff it needs from that. So it doesn't _really_ need an owned type. If that took a borrow, you could invert this call with the call above it and save yourself a potentially expensive copy